Intermittent Test Failures in paged_attention Example under tensormap_and_ringbuffer

## Problem

The `examples/tensormap_and_ringbuffer/paged_attention` test case exhibits non-deterministic failures when executed on A2/A3 hardware. When running 100 iterations in batch mode, approximately 6% of executions fail with output tensor mismatches against the golden reference.

The failures exhibit the following characteristics:
- **Failure rate**: Approximately 6% (6 failures out of 100 runs)
- **Non-deterministic behavior**: Failures occur at random run numbers (e.g., 19, 36, 47, 63, 89, 93) with no discernible pattern
- **Variable mismatch count**: The number of mismatched elements varies per failure, ranging from 5 to 24 out of 256 total output elements
- **Hardware-specific**: The issue only reproduces on actual hardware and appears to be timing-related

## Environment

- **Platform**: A2/A3 hardware (non-simulation)
- **Runtime**: tensormap_and_ringbuffer
- **Architecture**: ARM64 (Linux aarch64)
- **Test case path**: `examples/tensormap_and_ringbuffer/paged_attention/`

## Steps to Reproduce

1. Navigate to the simpler working directory:

```bash
cd simpler
```

2. Run the test case once (single runs usually pass, failures are intermittent):

```bash
python examples/scripts/run_example.py \
    -k examples/tensormap_and_ringbuffer/paged_attention/kernels \
    -g examples/tensormap_and_ringbuffer/paged_attention/golden.py \
    -p a2a3 -d=12
```

3. Use the batch script to reproduce (recommended - 100 runs reliably expose the issue):

```bash
bash pa_a2a3.sh
```

pa_a2a3.sh:
```bash
#!/bin/bash

TOTAL=100
PASS=0
FAIL=0

LOGFILE="batch_run_$(date '+%Y%m%d_%H%M%S')-pa_a2a3.txt"

CMD="python examples/scripts/run_example.py \
    -k examples/tensormap_and_ringbuffer/paged_attention/kernels \
    -g examples/tensormap_and_ringbuffer/paged_attention/golden.py \
    -p a2a3 -d=12"

echo "Log file: $LOGFILE"
echo "Batch run started at $(date)" | tee "$LOGFILE"
echo "" | tee -a "$LOGFILE"

for i in $(seq 1 $TOTAL); do
    echo "=== Run $i / $TOTAL ===" | tee -a "$LOGFILE"
    output=$($CMD 2>&1)
    echo "$output" >> "$LOGFILE"
    if echo "$output" | grep -q "TEST PASSED"; then
        PASS=$((PASS + 1))
        echo "[Run $i] PASSED" | tee -a "$LOGFILE"
    else
        FAIL=$((FAIL + 1))
        echo "[Run $i] FAILED" | tee -a "$LOGFILE"
    fi
    echo "" >> "$LOGFILE"
done

echo "" | tee -a "$LOGFILE"
echo "==============================" | tee -a "$LOGFILE"
echo "Results: $PASS / $TOTAL PASSED" | tee -a "$LOGFILE"
echo "Batch run finished at $(date)" | tee -a "$LOGFILE"
echo "==============================" | tee -a "$LOGFILE"
echo "Full log saved to: $LOGFILE"
```

This script executes 100 consecutive runs and tracks PASS/FAIL statistics.

## Error Output

When failures occur, `run_example.py` raises the following exception:

```
[ERROR] TEST FAILED: Output 'out' does not match golden.
Mismatched elements: <N>/256
rtol=0.01, atol=0.01
```

Where `<N>` varies across different failure instances. Typical values include:

| Failed Run Number | Mismatched Elements (/256) |
|-------------------|---------------------------|
| Run 19           | 5                         |
| Run 36           | 6                         |
| Run 47           | 9                         |
| Run 63           | 11                        |
| Run 89           | 15                        |
| Run 93           | 24                        |

Batch test summary output:

```
==============================
Results: 94 / 100 PASSED
Batch run finished at <timestamp>
==============================
```

## Expected Behavior

The test case should pass with **100% reliability**, i.e., batch runs should result in `100 / 100 PASSED`, with output tensors matching golden reference within the tolerance of `rtol=0.01, atol=0.01`.

## Additional Context

- The issue involves the `TFILLPAD_INPLACE` operation in `aiv_softmax_prepare.cpp`, specifically the subsequent `SetValue` writes to padding regions with -inf values
- Additionally, line 185 in `golden.py` uses `torch.bfloat16` to truncate pij, while the device-side kernel uses `float16` (`TCVT` fp32→fp16). This precision mismatch may introduce additional comparison errors

--- a/examples/tensormap_and_ringbuffer/paged_attention/golden.py
+++ b/examples/tensormap_and_ringbuffer/paged_attention/golden.py
@@ -179,7 +179,7 @@ def paged_attention(
             pij = torch.exp(sij - mij)
             pij = pij.masked_fill(~valid_mask, 0.0)
             pij = pij.masked_fill(~batch_mask, 0.0)
-            pij = pij.to(torch.bfloat16).to(torch.float32)
+            pij = pij.to(torch.float16).to(torch.float32)
             lij = pij.sum(dim=-1, keepdim=True)
 
             oi_new = torch.bmm(pij, vj_all)

--- a/examples/tensormap_and_ringbuffer/paged_attention/kernels/aiv/aiv_softmax_prepare.cpp
+++ b/examples/tensormap_and_ringbuffer/paged_attention/kernels/aiv/aiv_softmax_prepare.cpp
@@ -94,12 +94,25 @@ static __aicore__ void softmax_prepare_impl(__gm__ TensorData* sij,
     // Patch: SetValue ensures correctness for valid_len <= N/2 where
     // TFILLPAD's PadRightRemainingRows vcopy has a hardware issue.
     if (valid_len < static_cast<uint64_t>(N)) {
+        // Cross-pipeline sync: wait for PIPE_V vcopy in TFILLPAD to complete
+        // before PIPE_S scalar SetValue writes to the same UB addresses.
+        // Without this, PIPE_V vcopy and PIPE_S SetValue race on UB memory,
+        // causing sporadic FAIL when vcopy finishes after SetValue.
+        // Pattern from TFillPad.hpp Handle32BAlignedPad_Byte (PtoSetWaitFlag).
+        set_flag(PIPE_V, PIPE_S, EVENT_ID0);
+        wait_flag(PIPE_V, PIPE_S, EVENT_ID0);
         constexpr float NEG_INF = -__builtin_huge_valf();
         for (int r = 0; r < M; r++) {
             for (uint64_t c = valid_len; c < N; c++) {
                 sijTile.SetValue(static_cast<uint32_t>(r * N + c), NEG_INF);
             }
         }
+        // Ensure PIPE_S scalar UB writes are visible to subsequent PIPE_V ops.
+        // dsb(DSB_UB) is a hardware-only intrinsic; in simulation there are no
+        // real pipelines so the barrier is unnecessary and DSB_UB is undefined.
+#ifdef DSB_UB
+        dsb(DSB_UB);
+#endif
     }
 
     TMULS(sijTile, sijTile, scale_value);

